libs/wingc/src/type_check.rs

```diff
@@ -1188,6 +1240,146 @@ impl Types {
 		self.get_typeref(self.types.len() - 1)
 	}
 
+	/// If the type is inferred and the real type is known, return it.
+	/// Otherwise return the type itself.
+	///
+	/// Note: This function is recursive, so it will unwrap nested inferred types.
+	pub fn maybe_unwrap_inference(&self, t: TypeRef) -> TypeRef {
+		if let Type::Inferred(id) = &*t {
+			if let Some(t) = self.get_inference_by_id(*id) {
+				return self.maybe_unwrap_inference(t);
+			}
+		}
+
+		t
+	}
+
+	/// If the given type is inferred and the real type is known, return it
+	pub fn get_inference(&self, t: TypeRef) -> Option<TypeRef> {
+		if let Type::Inferred(id) = *t {
+			self.get_inference_by_id(id)
+		} else {
+			None
+		}
+	}
+
+	/// If the real type is known for the given inference id, return it
+	pub fn get_inference_by_id(&self, id: usize) -> Option<TypeRef> {
+		if let Some(t) = self.inferences.get(id).expect("Inference id out of bounds") {
+			return Some(*t);
+		}
+
+		None
+	}
+
+	pub fn update_inferred_type(&mut self, id: usize, t: TypeRef) {
+		let mut new_type = t;
+		if let Type::Inferred(n) = &*new_type {
+			if *n != id {
+				new_type = self.maybe_unwrap_inference(new_type);
+			}
+		}
+
+		let existing_type = self.inferences.get_mut(id).expect("Inference id out of bounds");
+
+		if existing_type.is_none() {
+			existing_type.replace(new_type);
+		} else {
+			// We already have a type for this inference
+			return;
+		}
+
+		for expr_type in self.type_for_expr.iter_mut().flatten() {
```

Comment:
I don't think fold is necessary since it's not part of the AST. However, I can look into a way to avoid this frequent iteration.

Reply:
Right, some visit pattern on types.

Reply:
Ok so for this particular section, this iteration actually isn't needed at all. I just deleted this section. The expression types get mutated elsewhere as needed so there's no needed to go through all the expression types preemptively.

Reply:
For a visitor pattern on types (for the implementation of `deep_update_inference`), I don't think that would help perf but it would probably look better. Should I tackle that as part of this PR?

Reply:
I do think it'll look better and might reduce future bugs. If you feel like it then It'll be good for it to go in now. But not critical if it turns out complex.

Reply:
Alighty, added a visitor for types now